### ISSUE TITLE
ControlStore issues

### ISSUE THREAD

**Issue by timelyportfolio** [rstudio/leaflet#401]
I discovered that many/all of the old `controls` do not use the implemented [`ControlStore`](https://github.com/rstudio/leaflet/blob/master/javascript/src/control-store.js) mechanism.  I am happy to fix this and will submit a pull.

However, I believe the `add`/`remove` mechanisms are incomplete, since `controls` without `id` are added but cannot be removed with `ControlStore`.  In these [lines](https://github.com/rstudio/leaflet/blob/master/javascript/src/control-store.js#L22-L26), only controls with initial `id` can be removed.  I believe since `leaflet` stamps all controls, we should in the add method also `push` the no-id `control` to `this._controlsById` once leaflet has stamped it.  Then, the `remove` method will work as expected.

**Comment by timelyportfolio** [comment]
Never mind, I see my error.  `addControls` is the way to do this.

**Comment by bhaskarvk** [comment]
Btw, addControl requires you to input HTML code which it wraps in a `div`.
But if you already have a Control Object, the function to call is `addCustomControl` https://github.com/rstudio/leaflet/blob/master/javascript/src/methods.js#L671`.
I use that quite a bit in leaflet.extras, and measure, and other controls should be added using it.

**Comment by timelyportfolio** [comment]
Actually, as I dig deeper, it seems none of the controls use `methods.addControl` or `methods.addCustomControl`.  Rather, `this.addControl` is simply the `addControl` for `leaflet`.  So the `ControlStore` mechanism is being bypassed completely.

@bhaskarvk, I could not find any examples of `addCustomControl` in `leaflet-extras`.  Can you point me to a line?

**Comment by bhaskarvk** [comment]
Yikes,
I think you are right. leaflet.extras doesn't use controlstore, and I think I should change that before I submit it to cran.
Also looking at leaflet code, it seems the only thing that's using the control store as of now is addLegend.

**Comment by timelyportfolio** [comment]
Will check out addLegend since I did not see that it uses ControlStore.  Using ControlStore through methods.*Control is kind of clunky, and there is no way to remove if no id assigned.  I plan to experiment tomorrow with some small changes that I think will greatly help.

**Comment by jcheng5** [comment]
@timelyportfolio What are the built-in control functions that ship with Leaflet? `addLayersControl`, `addLegend`, and `addControl`? (easy button?)

**Comment by timelyportfolio** [comment]
Off computer now but the ones I remember seeing are layersControl, measure (maybe should move out of leaflet core), selectButton, legend.  Ones that are currently broken are layers and measure.  I fixed to "work" in my v1.0 branch.  I will assemble an entire list and some ideas for improvement tomorrow.

**Comment by jcheng5** [comment]
I don't love the idea of JavaScript callers calling methods.* methods--those were originally intended just to service R calls. Though one thing led to another with the refactor of the various types of markers, I suppose. It would be better if the reusable JavaScript methods were accessed via a different object to make it clear what is intended for R interop and what is intended to be called by JS.

**Comment by bhaskarvk** [comment]
I agree with @jcheng5, `methods.*` should be strictly for R APIs to call.
FWIW, we have some code that looks like

```javascript

function addLayer(....)

methods.addLayer = function(...) {
   addLayer(...)
}

methods.addWMSLayer = function(...) {
  addLayer(...)
}

methods.addGenericLayer = addLayer
```

So a generic addLayer function, that gets called by `methods` function, and also exported as `addGenericLayer` for use in downstream packages like `leaflet.extras` and `leaflet.esri`.

Similarly we have `methods.addGenericMarkers` and `addCustomControl`.

I think that is a nice compromise. But always willing to consider better alternatives if any

**Comment by timelyportfolio** [comment]
I agree that the JS and R methods/functions should be separate.  That will solve the "clunkiness" I mentioned, since using the `methods.*` require

```
methods.*.call(this, ...)
```

which works but is a pain.

I guess I am troubled slightly though since `addControl` and `removeControl` are already available at the `L.*` level, so maybe a different name like `addCustomControl` and `removeCustomControl` would be less confusing.

Also,  I very much like being able to supply the control object to `add/remove` instead of a `layerId` which might or might not exist. Any thoughts, opinions on this topic?

**Comment by jcheng5** [comment]
Keep in mind that the add/remove based on layerId is designed for use from R, not JavaScript. In fact AFAICT methods.addControl/methods.removeControl are exclusively designed for use from R?

**Comment by timelyportfolio** [comment]
Yes and understood completely.  However, it seems `ControlStore` is designed to be the interface for both R & JS, so I think I am leaning toward additional functions or additional arguments within `ControlStore`.

**Comment by bhaskarvk** [comment]
@jcheng5 For extensibility (think `leaflet.extras` and more so `leaflet.esri`) I need to use these methods not only from R but also from JS. 
I have taken care not to call `methods.*` directly from any JS code in the above pkgs, but I do need to use the JS functions and objects that the `methods.add*` functions use internally such as `controls` (ControlStore), `layerManager`, etc.

This issue is not specific to `leaflet` per se, this is something any widget that needs extensibility will have to deal with. How to provide R/JS API/Objects that can be used properly by extending packages. For now looks like apart from `leaflet` no other widget is really extensible, but doesn't hurt to think of a problem with a broader scope. :)

**Comment by timelyportfolio** [comment]
I think @daattali did a pretty good job with this in `timevis`.  See https://github.com/daattali/timevis/blob/master/inst/htmlwidgets/timevis.js.

**Comment by daattali** [comment]
If the topic being discussed is how to provide an API callable by big R and
javascript in an htmlwidget, then perhaps my solution can be useful. I
wouldn't be surprised if you find something strange in my code since it was
my first widget attempt, but in case it helps I wrote a bit about the
approach I chose http://deanattali.com/blog/htmlwidgets-tips/#api

If this was irrelevant, then feel free to ignore :)
